source/connections-survive-step-down/tests/README.rst

```diff
@@ -0,0 +1,214 @@
+the case in the getMore test for example, the driver should perform the
+following procedure:
+
+- Obtain a list of servers in the cluster.
+- Randomly or otherwise choose a server to be the new primary, other than the
+  current primary.
+- Follow the above server election procedure to elect the chosen server as
+  the new primary.
+
+Tests
+-----
+
+The driver should implement the following tests:
+
+getMore Iteration
+`````````````````
+
+This test requires server version 4.2 or higher.
+
+Perform the following operations:
+
+- Insert 5 documents into a collection.
+- Start a find operation on the collection with a batch size of 2, and
+  retrieve the first batch of results.
+- Step down the primary, and verify that another server has been elected as
+  a primary.
+- Using the same connection that was used for the initial find, continue
+  iterating the collection by issuing a getMore operation.
+- Verify that the next batch of results is returned successfully.
+
+Drivers that do not ordinarily allow applications to execute operations
+on specific connections may need to perform the getMore iteration test on
+a lower level than the public API exposed to applications.
+In this case the driver SHOULD verify that both initial find and the
+subsequent getMore are dispatched on the same connection, for example by
+comparing socket handles used for the two operations.
+
+Acknowledged Write
+``````````````````
+
+This test requires server version 4.2 or higher.
+
+Perform the following operations:
+
+- Obtain a connection to the current primary.
+- Step down the primary, and verify that another server has been elected as
+  a primary.
+- Dispatch an acknowledged insert operation over the connection obtained in
+  the first step.
+- Verify that the result of the insert operation is a NotMaster error.
+- Verify that the connection used for the insert operation remains open.
+
+Unacknowledged Write
+````````````````````
+
+This test requires server version 4.2 or higher.
+
+Perform the following operations:
+
+- Obtain a connection to the current primary.
+- Step down the primary, and verify that another server has been elected as
+  a primary.
+- Dispatch an unacknowledged insert operation over the connection obtained
+  in the first step.
+- Dispatch another unacknowledged insert operation over the connection obtained
+  in the first step.
+- If the second unacknowledged insert succeeded, dispatch another unacknowledged
+  insert over the same connection[*].
+- Verify that the last unacknowledged insert produced a network error due
+  to the connection being closed by the server.
+
+[*] Depending on buffering performed by the network stack, the first write
+operation on a connection which was closed on the remote end may succeed.
+In the acknowledged write scenario, the error is detected upon (trying to)
+read the response. In the unacknowledged write scenario, an additional write
+may be needed to detect that the connection is closed.
```

Comment:
This talks about an additional write being needed to detect a closed connection; however, the sequence of operations suggests three possible inserts after the step down. Why can't we trust the second insert to detect the closed connection?

Reply:
The second insert is actually the first insert after the close, and since the close is on the other end only (our end hasn't read or written anything yet and thus has no way of being aware of the close), the first write simply buffers the data and succeeds. I needed to write over 20k to the connection to have the first write fail.

I also ended up putting a half second sleep between the second and the third insert because sometimes even two consecutive inserts would be successfully buffered and not error.

Reply:
That explanation makes sense. If the server closes the connection after the first unacknowledged write, the earliest a driver might detect a closed socket is when attempting to write the second insert. And you're explaining that we might need another attempt (second after the initial closure, but third overall) to overcome buffering or other subtleties.

It may be worth explaining this in a bit more detail for the benefit of others reading the final prose test. Also, a tip about using larger insert payloads to overcome network buffering might be helpful. IIRC, 8 KiB is a common network buffer size but performing an insert with 64+ KiB of data is probably safer.

Reply:
Would you prefer the language requiring the driver to send a sufficiently large payload (64kb+ as you said) to overcome network buffering over the current language suggesting multiple writes?

Reply:
Are two unacknowledged writes always guaranteed to overcome network buffering? If the buffer is 8 KiB, I'm curious how two very small writes might fill that up -- unless there is some other factor responsible for flushing the buffer.

Reply:
They are not, I put a sleep between them to force the issue.

Reply:
Ah, in that case I'd vote for using larger writes in lieu of a sleep. Happy to defer to someone else's opinion on this, though.